resolver-dns/src/main/java/io/netty/resolver/dns/AuthoritativeDnsServerCache.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns;
+
+import io.netty.channel.EventLoop;
+import io.netty.util.internal.UnstableApi;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+/**
+ * Cache which stores the nameservers that should be used to resolve a specific hostname.
+ */
+@UnstableApi
+public interface AuthoritativeDnsServerCache {
+
+    /**
+     * Returns the cached nameservers that should be used to resolve the given hostname. The returned unmodifiable
+     * {@link List} may contain unresolved {@link InetSocketAddress}es that will be resolved when needed while
+     * resolving other domain names.
+     *
+     * @param hostname the hostname
+     * @return the cached entries
+     */
+    List<InetSocketAddress> get(String hostname);
+
+    /**
+     * Caches a nameserver that should be used to resolve the given hostname.
+     *
+     * @param hostname the hostname
+     * @param address the nameserver address (which may be unresolved).
+     * @param originalTtl the TTL as returned by the DNS server
+     * @param loop the {@link EventLoop} used to register the TTL timeout
+     */
+    void cache(String hostname, InetSocketAddress address, long originalTtl, EventLoop loop);
```

Comment:
This API is odd, because DNS always gives you back lots of addresses at the same TTL, not one.  If you had to call this in a loop, the TTLs might slide.  This would be surprising as a user.  

Consider taking a list of records for a given TTL.

Reply:
@carl-mastrangelo while the TTL may be the same its not required...

```
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 62895
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 9
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;ns1.google.com.			IN	A

;; AUTHORITY SECTION:
google.com.		172800	IN	NS	ns2.google.com.
google.com.		172800	IN	NS	ns1.google.com.
google.com.		172800	IN	NS	ns3.google.com.
google.com.		172800	IN	NS	ns4.google.com.

;; ADDITIONAL SECTION:
ns2.google.com.		172800	IN	AAAA	2001:4860:4802:34::a
ns2.google.com.		172800	IN	A	216.239.34.10
ns1.google.com.		172800	IN	AAAA	2001:4860:4802:32::a
ns1.google.com.		172800	IN	A	216.239.32.10
ns3.google.com.		172800	IN	AAAA	2001:4860:4802:36::a
ns3.google.com.		172800	IN	A	216.239.36.10
ns4.google.com.		172800	IN	AAAA	2001:4860:4802:38::a
ns4.google.com.		172800	IN	A	216.239.38.10
```

As you can see each of these define their own TTL. Sure its the same here but this is not required by the protocol so we should not make assumptions here.

Reply:
In the case the TTL is the same, I am suggesting that they be added atomically.   If the cache is queried right as the scheduled removal is happening, the cache may return partial results.  This is what I meant by sliding.

Also, not clear on the expected performance needs here, but I would also assume a single scheduled future is lighter weight than four.

Reply:
+1 for consolidating cache records but perhaps we could take care of it in a follow-up PR to fix both `DnsCache` and `AuthoritativeDnsServerCache`?

Reply:
At least in our current implementation it’s not a problem as we just remove all entries once the smallest TTL matches. This may result in more queries then needed but is the safer bet to not end up with partial results imho:

https://github.com/netty/netty/pull/8177/files#diff-7ab1cd3db010ccc37c01987e522675caR178

That said I guess we could also just schedule for the smallest TTL here. 

WDYT?

Reply:
Here is a PR for the "smallest TTL change" for the `DefaultDnsCache`:

https://github.com/netty/netty/pull/8187/

Something like this could also be used for the `DefaultAuthoritativeDnsServerCache` here. Let me know what you think @carl-mastrangelo @trustin

Reply:
Sounds good to me.